src/main/java/net/sf/jabref/model/database/KeyChangeListener.java

```diff
@@ -0,0 +1,78 @@
+package net.sf.jabref.model.database;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import net.sf.jabref.model.entry.BibEntry;
+import net.sf.jabref.model.entry.FieldProperties;
+import net.sf.jabref.model.entry.InternalBibtexFields;
+import net.sf.jabref.model.event.EntryRemovedEvent;
+import net.sf.jabref.model.event.FieldChangedEvent;
+
+import com.google.common.eventbus.Subscribe;
+
+public class KeyChangeListener {
+
+    private final BibDatabase database;
+
+    private final List<String> keyFields = new ArrayList<>();
+
+
+    public KeyChangeListener(BibDatabase database) {
+        this.database = database;
+
+        // Look for fields with FieldProperies.SINGLE_ENTRY_LINK or FieldProperties.MULTIPLE_ENTRY_LINK to speed up the search later
+
+        for (String fieldName : InternalBibtexFields.getAllFieldNames()) {
+            if (InternalBibtexFields.getFieldExtras(fieldName).contains(FieldProperties.SINGLE_ENTRY_LINK)
+                    || InternalBibtexFields.getFieldExtras(fieldName).contains(FieldProperties.MULTIPLE_ENTRY_LINK)) {
+                keyFields.add(fieldName);
+            }
+        }
+    }
+
+    @Subscribe
+    public void listen(FieldChangedEvent event) {
+        if (event.getFieldName().equals(BibEntry.KEY_FIELD)) {
+            String newKey = event.getNewValue();
+            String oldKey = event.getOldValue();
+            updateEntryLinks(newKey, oldKey);
+        }
+    }
+
+    @Subscribe
+    public void listen(EntryRemovedEvent event) {
+        String oldKey = event.getBibEntry().getCiteKey();
+        updateEntryLinks(null, oldKey);
+    }
+
+    private void updateEntryLinks(String newKey, String oldKey) {
+        for (BibEntry entry : database.getEntries()) {
+            for (String field : keyFields) {
+                entry.getFieldOptional(field).ifPresent(fieldContent -> {
+                    if (InternalBibtexFields.getFieldExtras(field).contains(FieldProperties.SINGLE_ENTRY_LINK)) {
+                        if (fieldContent.equals(oldKey)) {
```

Comment:
Maybe move the contents of the if and else bodies to seperate methods to increase readability.